src/adapter/src/coord/message_handler.rs

```diff
@@ -135,9 +142,13 @@ impl<S: Append + 'static> Coordinator<S> {
             .storage_usage_collection_interval
             .saturating_sub(Duration::from_millis(time_since_previous_collection));
         let internal_cmd_tx = self.internal_cmd_tx.clone();
+        let now_fn = self.now_fn();
         task::spawn(|| "storage_usage_collection", async move {
             tokio::time::sleep(next_collection_interval).await;
-            if internal_cmd_tx.send(Message::StorageUsageFetch).is_err() {
+            if internal_cmd_tx
+                .send(Message::StorageUsageFetch(now_fn()))
```

Comment:
I do not think this will change the behavior to do what you describe in the comment. Current we:

1. Wait for an hour.
2. Request storage size.
3. Wait an unknown amount of time.
4. Get storage size.
5. Determine a timestamp.
6. Send the data..
7. Go to step 1.

The current diff here moves step 5 to before step 2, which means the times between measurements will still drift apart. My read of your commit message is that what you want is: change step 7 to be:

> Wait for (1 hour - the time from step 3). Go to step 2.

If that's a correct summary, the fix for this is to change the `::sleep` to a `tick`, which always produces an event each interval, and thus doesn't care how long step 3 took.

Reply:
I think that step 1 actually is wait for (1 hour - the time from step 3). The timestamp from step 3 is saved in `most_recent_storage_usage_collection` here: https://github.com/MaterializeInc/materialize/blob/f0fa147d54213d359acebd787b174561840df7f8/src/adapter/src/coord/message_handler.rs#L125-L126

Then when determining how long to wait we do `storage_usage_collection_interval - (now - `most_recent_storage_usage_collection)` here: https://github.com/MaterializeInc/materialize/blob/f0fa147d54213d359acebd787b174561840df7f8/src/adapter/src/coord/message_handler.rs#L130-L144

The reason we don't use an interval is because we want to try and keep consistent intervals in-between restarts. I should probably add a comment with this to `schedule_storage_usage_collection`.

Reply:
Oop yes, that sounds correct!